### ISSUE TITLE
[css] css language server gets diagnistic pull requests for typescript files

### ISSUE THREAD

**Issue by aeschli** [microsoft/vscode#180152]
Version: 1.78.0-insider (user setup)
Commit: c9df8ee8e8c316f38dcc75a894ba150024ce3718
Date: 2023-04-13T05:25:44.823Z
Electron: 22.4.4
Chromium: 108.0.5359.215
Node.js: 16.17.1
V8: 10.8.168.25-electron.0
OS: Windows_NT x64 10.0.22624
Sandboxed: Yes

- enable `"css.trace.server": "verbose",`
- open a css file and open the `CSS Language Server` output
- now open several typescript files and observe that for each opened file there's diagnostic request made against the CSS server

**Comment by dbaeumer** [comment]
I debugged it and the reason for this is that when I receive the tabs open event that active editor and the text document state has not been updated yet. I am pretty sure that this was different in the past. 

In the attached screen shot the debugger pauses on the `window.tabGroups.onDidChangeTabs` event. Here is what I did:

- had codeEditor.ts open in the editor
- clicked on contextkeys.ts

The active editor clearly shows `contextkeys.ts` however the active editor is still:
```
vscode_1.window.activeTextEditor.document.uri.toString()
'file:///home/dirkb/Projects/mseng/VSCode/vscode/src/vs/workbench/browser/codeeditor.ts'
```

In addition `workspace.textDocuments` still list the codeeditor as well

```
vscode_1.workspace.textDocuments[3].uri.toString()
'file:///home/dirkb/Projects/mseng/VSCode/vscode/src/vs/workbench/browser/codeeditor.ts'
```
<img width="1601" alt="image" src="https://user-images.githubusercontent.com/1931590/232702159-40277f70-2ec3-4822-afde-93d3533c8a46.png">

**Comment by dbaeumer** [comment]
@lramos15 is this now expected. I think if the editor shows the new file the tab events when fired should allow me to access the new information and not the old one.

**Comment by dbaeumer** [comment]
@aeschli you could for now work around this by adding a match function to the `DiagnosticPullOptions`.

**Comment by dbaeumer** [comment]
I think we should at least specify the behavior and have a test case. Otherwise it is very hard to program against these.

**Comment by lramos15** [comment]
@bpasero Assigning you as well as this seems to happen when the tabs API listener fires before the active editor is ready because the tab has updated? Do you have any ideas if this is a tabs bug or active editor bug.

Here's a simple extension repro.

```
	const tabListener = vscode.window.tabGroups.onDidChangeTabs(() => {
		console.log('Tab Listener');
		console.log(`Active Tab: ${vscode.window.tabGroups.activeTabGroup.activeTab.label}`);
		console.log(`Active Editor: ${vscode.window.activeTextEditor.document.fileName}`);
	});

	const editorListener = vscode.window.onDidChangeActiveTextEditor(() => {
		console.log('Editor Listener');
		console.log(`Active Tab: ${vscode.window.tabGroups.activeTabGroup.activeTab.label}`);
		console.log(`Active Editor: ${vscode.window.activeTextEditor.document.fileName}`);
	});
```

Switching a tab you will see tabs fires first with the tab having the new label but the active editor having the old label. Then the activeTextEditor listener fires with the active editor being the correct new value

**Comment by bpasero** [comment]
@lramos15 that is expected, that event only fires when the input of the editor has resolved:

https://github.com/microsoft/vscode/blob/34b34c4577ca4ce67fa22de3eec626de32da60e1/src/vs/workbench/browser/parts/editor/editorGroupView.ts#L1075-L1085

I would always expect the tabs event to fire first because opening a tab is instant.

**Comment by lramos15** [comment]
Yeah that was my hypothesis as well, so it seems both things are working properly here. Is there a listener that @dbaeumer can use to know when the document list has changed besides just the active one?

**Comment by dbaeumer** [comment]
I looked at my code and I can code it either way however it needs to be document how these events fire:

- I can rely on the document open event. But since I only want to sync visible documents I needs to be guaranteed that there is a tab representing the editor available via the API and that the tab event has already fired.
- Or I rely on the tabs events but then it needs to be guaranteed that the document / active editor is already updated.

Let me know which why we would specify it and I will adjust my code accordingly

**Comment by dbaeumer** [comment]
@aeschli 



I found another easier workaround for you to avoid this: change your document selector in CSS extension to 
```
[
  { language: 'css' },
  { language: 'scss'}
  { language: 'less'}
]
```

**Comment by lramos15** [comment]
> * I can rely on the document open event. But since I only want to sync visible documents I needs to be guaranteed that there is a tab representing the editor available via the API and that the tab event has already fired.

I think this is the correct one. We always expect a tab to be ready before its backing editor is there. Therefore if the backing editor is there then the tab should be too.

**Comment by dbaeumer** [comment]
Moving to vscode-languageserver-node
